Sources/Sentry/SentryFramesTracker.m

```diff
@@ -134,28 +134,26 @@ - (void)displayLinkCallback
     }
 #    endif // SENTRY_TARGET_PROFILING_SUPPORTED
 
-    // Most frames take just a few microseconds longer than the optimal caculated duration.
-    // Therefore we substract one, because otherwise almost all frames would be slow.
+    // Most frames take just a few microseconds longer than the optimal calculated duration.
+    // Therefore we subtract one, because otherwise almost all frames would be slow.
     CFTimeInterval slowFrameThreshold = 1 / (actualFramesPerSecond - 1);
 
-    CFTimeInterval frameDuration = lastFrameTimestamp - self.previousFrameTimestamp;
-    self.previousFrameTimestamp = lastFrameTimestamp;
+    CFTimeInterval frameDuration = renderedFrameTimestamp - self.previousFrameTimestamp;
 
     if (frameDuration > slowFrameThreshold && frameDuration <= SentryFrozenFrameThreshold) {
         atomic_fetch_add_explicit(&_slowFrames, 1, SentryFramesMemoryOrder);
 #    if SENTRY_TARGET_PROFILING_SUPPORTED
-        [self recordTimestampStart:@(self.previousFrameTimestamp) end:@(lastFrameTimestamp)];
+        [self recordTimestampStart:@(self.previousFrameTimestamp) end:@(renderedFrameTimestamp)];
 #    endif // SENTRY_TARGET_PROFILING_SUPPORTED
     } else if (frameDuration > SentryFrozenFrameThreshold) {
         atomic_fetch_add_explicit(&_frozenFrames, 1, SentryFramesMemoryOrder);
 #    if SENTRY_TARGET_PROFILING_SUPPORTED
-        [self recordTimestampStart:@(self.previousFrameTimestamp) end:@(lastFrameTimestamp)];
+        [self recordTimestampStart:@(self.previousFrameTimestamp) end:@(renderedFrameTimestamp)];
 #    endif // SENTRY_TARGET_PROFILING_SUPPORTED
     }
 
     atomic_fetch_add_explicit(&_totalFrames, 1, SentryFramesMemoryOrder);
-
-    self.previousFrameTimestamp = lastFrameTimestamp;
+    self.previousFrameTimestamp = renderedFrameTimestamp;
```

Comment:
This was the problem. The same assignment to `self.previousFrameTimestamp` appeared both before and after recording the timestamps. It should only have been at the end.

Reply:
I noticed the same problem in one of my PRs but that got closed without getting merged. Should've copied the fix into a new PR 😓

But the fix seemed like a one-liner to me? See https://github.com/getsentry/sentry-cocoa/pull/2028/files#r934466813.

Reply:
@armcknight, can you please add a test for this to avoid future bugs?

Reply:
@kevinrenskers You're right, the only line needed to fix the bug was removing the assignment on line 142. The rest is code cleanup: renaming variables, preventing divide-by-0, etc.